From e7cfd1ca763a1db7a222a07ffb077cc0067cba05 Mon Sep 17 00:00:00 2001
From: Nicolas Dufresne <nicolas.dufresne@collabora.com>
Date: Tue, 14 Apr 2020 17:10:34 -0400
Subject: [PATCH 34/60] kmssink: Save last metadata at the same time as the
 last buffer

The render width/height and the vinfo was only saved upon renegotiation. This
fixes the problem by saving this metadata at the same time the buffer is
saved. The saved copy of this is needed for expose() and drain() virtual functions.
This fixes various assertion that happens on drain query.

(cherry picked from commit 63940feaafa31c584a99fee006cae4f7b22146ed)
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 sys/kms/gstkmssink.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/sys/kms/gstkmssink.c b/sys/kms/gstkmssink.c
index d2c82df04..b32fd841d 100644
--- a/sys/kms/gstkmssink.c
+++ b/sys/kms/gstkmssink.c
@@ -1118,10 +1118,6 @@ gst_kms_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
 
   if (!gst_video_info_from_caps (&vinfo, caps))
     goto invalid_format;
-
-  self->last_width = GST_VIDEO_SINK_WIDTH (self);
-  self->last_height = GST_VIDEO_SINK_HEIGHT (self);
-  self->last_vinfo = self->vinfo;
   self->vinfo = vinfo;
 
   if (!gst_kms_sink_calculate_display_ratio (self, &vinfo,
@@ -1671,8 +1667,13 @@ sync_frame:
     goto bail;
   }
 
-  if (buffer != self->last_buffer)
+  /* Save the rendered buffer and its metadata in case a redraw is needed */
+  if (buffer != self->last_buffer) {
     gst_buffer_replace (&self->last_buffer, buffer);
+    self->last_width = GST_VIDEO_SINK_WIDTH (self);
+    self->last_height = GST_VIDEO_SINK_HEIGHT (self);
+    self->last_vinfo = self->vinfo;
+  }
   g_clear_pointer (&self->tmp_kmsmem, gst_memory_unref);
 
   GST_OBJECT_UNLOCK (self);
-- 
2.20.1

